lib/kino/smart_cell.ex

```diff
@@ -160,6 +160,34 @@ defmodule Kino.SmartCell do
   """
   @callback to_source(attrs()) :: String.t()
 
+  @doc """
+  Invoked whenever the base evaluation context changes.
+
+  This callback receives the binding and environment available to the
+  smart cell code.
+
+  Note that this callback runs asynchronously and the result is then
+  handled with `c:scan_binding_continue/2` where the server state can
+  be updated accordingly.
+
+  **Important:** the return value is copied between processes, so
+  avoid returning large data structures. In particular, when looking
+  at variables, instead of returning their values, extract and return
+  only the relevant metadata.
+
+  **Important:** avoid any heavy work in this callback, as it runs in
+  the same process that evaluates code, so we don't want to block it.
+  """
+  @callback scan_binding(Code.binding(), Macro.Env.t()) :: data :: term()
+
+  @doc """
+  Invoked with the result of `c:scan_binding/2` to update the server
+  state.
+  """
+  @callback scan_binding_continue(data :: term(), ctx :: Context.t()) :: ctx :: Context.t()
```

Comment:
For completeness, there is another API you can use here:

    scan_binding(pid, code, env)

Where `pid` is the `pid` of the smart cell. This way we have fewer callbacks and give a bit more flexibility to the user. I think I prefer this approach.

If you prefer two callbacks, then i would perhaps change the names to: `prepare_scan_binding` and `handle_scan_binding`. `handle_` is a known name for server callbacks after all and that would make it more apparent.